src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -49,14 +51,31 @@ public void discoverClusterNodesAndSlots(Jedis jedis) {
       this.nodes.clear();
       this.slots.clear();
 
-      String localNodes = jedis.clusterNodes();
-      for (String nodeInfo : localNodes.split("\n")) {
-        ClusterNodeInformation clusterNodeInfo = nodeInfoParser.parse(nodeInfo, new HostAndPort(
-            jedis.getClient().getHost(), jedis.getClient().getPort()));
+      List<Object> slots = jedis.clusterSlots();
 
-        HostAndPort targetNode = clusterNodeInfo.getNode();
-        setNodeIfNotExist(targetNode);
-        assignSlotsToNode(clusterNodeInfo.getAvailableSlots(), targetNode);
+      for (Object slotInfoObj : slots) {
+        List<Object> slotInfo = (List<Object>) slotInfoObj;
+
+        if (slotInfo.size() <= 2) {
+          continue;
+        }
+
+        List<Integer> slotNums = getAssignedSlotArray(slotInfo);
+
+        // hostInfos
+        int size = slotInfo.size();
+        for (int i = MASTER_NODE_INDEX; i < size; i++) {
```

Comment:
@marcosnils 
I thought it would be safer to discover as many nodes as we can, even though they're slaves because of failover.

Let's assume that Jedis knows about only masters.
Let's think about failover situation: we have 3 masters and 3 slaves, and we didn't notice that 3 masters are dead.
In a normal situation, 3 slaves should be promoted to master so there's no issue from Redis Cluster. But Jedis can't access any random nodes for redirection because all nodes Jedis knows are dead.